src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs

```diff
@@ -0,0 +1,84 @@
+
+using System;
+using System.Linq;
+using System.Linq.Expressions;
+using Microsoft.AspNetCore.Analyzers.RouteEmbeddedLanguage.Infrastructure;
+using Microsoft.AspNetCore.Analyzers.RouteEmbeddedLanguage.Infrastructure.VirtualChars;
+using Microsoft.AspNetCore.Analyzers.RouteEmbeddedLanguage.RoutePattern;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Analyzers.RouteHandlers;
+
+public partial class RouteHandlerAnalyzer : DiagnosticAnalyzer
+{
+    private static void DisallowNonParsableComplexTypesOnParameters(
+        in OperationAnalysisContext context,
+        WellKnownTypes wellKnownTypes,
+        IInvocationOperation invocation,
+        IMethodSymbol methodSymbol)
+    {
+        var routePatternArgument = invocation.Arguments[1];
+
+        if (routePatternArgument.Value.ConstantValue is not { HasValue: true } constant || constant.Value is not string routeTemplate)
+        {
+            return;
+        }
+
+        var enumerator = new RouteTokenEnumerator(routeTemplate);
+
+        while (enumerator.MoveNext())
+        {
+            var routeParameterName = enumerator.CurrentName.ToString();
+
+            var violatingParameter = methodSymbol.Parameters.FirstOrDefault(
+                p => p.Name == routeParameterName && !IsParsable(p.Type, wellKnownTypes)
+                );
+
+            if (violatingParameter != null)
+            {
+                var syntax = (ParameterSyntax)violatingParameter.DeclaringSyntaxReferences[0].GetSyntax(context.CancellationToken);
+                context.ReportDiagnostic(Diagnostic.Create(
+                    DiagnosticDescriptors.RouteParameterComplexTypeIsNotParsable,
+                    syntax.GetLocation(),
+                    routeParameterName,
+                    violatingParameter.Type.Name
+                    ));
+            }
+        }
+    }
+
+    private static bool IsParsableViaIParsable(ITypeSymbol typeSymbol)
+    {
+        var implementsIParsable = typeSymbol.AllInterfaces.Any(
+            i => i.Name == "IParsable" && // ... implements IParsable
+                 i.TypeArguments.Length == 1 && // ... being pedantic that it has just one type argument.
+                 i.TypeArguments.All(t => t.Name == typeSymbol.Name)); // ... and the type argument is the same as the type symbol
+
+        return implementsIParsable;
+    }
+
+    private static bool IsParsableViaTryParse(ITypeSymbol typeSymbol, WellKnownTypes wellKnownTypes)
+    {
+        var implementsTryParseMethod = typeSymbol.GetMembers().OfType<IMethodSymbol>().Any(m =>
+        {
+            return m.Name == "TryParse" &&
+                wellKnownTypes.Boolean.IsAssignableFrom(m.ReturnType) &&
+                wellKnownTypes.String.IsAssignableFrom(m.Parameters[0].Type) &&
```

Comment:
You don't need to do this for some built-in types. Test `SpecialType` for boolean and string.

See https://github.com/dotnet/aspnetcore/blob/4ff55988a7f634aa443693bdd917488b9b018cd9/src/Framework/AspNetCoreAnalyzers/src/Analyzers/Http/HeaderDictionaryIndexerAnalyzer.cs#L30